### PR TITLE
treewide: `set -u` everywhere

### PR DESCRIPTION
**Pull request by Ericson2314**
2 years or so ago I made stdenv and cc-wrapper use `set -u`, which *enormously* helped with debugging. But I didn't want to change all of nixpkgs so I added lots of temporary `set +u`-ing. 

Well, I think it's now time to bring the benefit to all of nixpkgs. Why? Because we have another big repo-wide refactor going on https://github.com/NixOS/nixpkgs/pull/72074. This will make that much easier to debug, and hopefully get some momentum going for stdenv tech debt fixing. Everyone chip in!

In particular, I think this is the easiest stdenv refactor we'd like to make, Why? This has the nice property that programs that continue to not fail have the same behavior. [Actually, this is  not true with `trap` to catch failures or no `set -e`, but we use `set -e` and still fail after `trap`-ing.] That means it's very easy to debug; no getting a bunch of things working only to find some deeper dependency was built wrong and needing to backtrack, and no needing to debug complex failures.

I swear there was an issue for thus, but I couldn't find it.

Hydra job: https://hydra.nixos.org/jobset/nixpkgs/bash-no-undef-vars#tabs-evaluations

TODO:

- [x] Fix the broken things

- [x] Write release notes

### CODE REVIEW

**Comment by FRidh** [comment]
How likely is it to break things and how much as been tested?

**Comment by Mic92** [comment]
This should be also mentioned in the release notes as it can break out-of-tree nixpkgs.

**Comment by Ericson2314** [comment]
Wow, is there no way to mark a PR draft after the fact?!?! Embarrising, GitHub.

**Comment by Ericson2314** [comment]
Hydra reached the point where the legitimate failures were buried amidst all sorts of spurious failures. This makes it harder to use it. I think it might be ready to go to staging then, at least after this these newer evals get further? @globin @matthewbauer

**Comment by Ericson2314** [comment]
https://hydra.nixos.org/eval/1552118?compare=1551936&full=1#tabs-now-fail here is a comparison to the base commit, which is a nixpkgs-unstable eval.

File: pkgs/build-support/setup-hooks/multiple-outputs.sh [hunk 1]
```diff
@@ -9,8 +9,8 @@ _assignFirst() {
     local varName="$1"
     local REMOVE=REMOVE # slightly hacky - we allow REMOVE (i.e. not a variable name)
     shift
-    while [ $# -ge 1 ]; do
-        if [ -n "${!1}" ]; then eval "${varName}"="$1"; return; fi
+	while (( $# )); do
```
Comment (matthewbauer) [inline pkgs/build-support/setup-hooks/multiple-outputs.sh]: indentation

**Review by matthewbauer — APPROVED** [review]

**Comment by Ericson2314** [comment]
@FRidh that is a commit pushed post merge.

**Comment by Mic92** [comment]
qt needs a fix: https://github.com/NixOS/nixpkgs/pull/73586#issuecomment-555380665

**Comment by FRidh** [comment]
qtWrapperArgs is fixed with ef43c5f5d928758d15cd42b18454e7145a163331. Now there are issues left with the setup hooks of both plasma5 packages and kdeFrameworks packages.
`pkgs/desktops/plasma-5/default.nix`
`pkgs/development/libraries/kde-frameworks/default.nix`
Looked into this briefly, https://github.com/NixOS/nixpkgs/pull/73756

**Comment by Ericson2314** [comment]
Actually my commit could not be enough, but @mic92's won't work either. The issue is bash says an undefined variable and and undefined variable are the same thing (which is terrible!!!).

The direct workaround is awful `${foo+${foo[@]}}`, but thankfully there are other ways like `${#foo[@]}` to get the same result.

**Comment by jtojnar** [comment]
Hmm, it appears this is not enough. For example `wrapPython` setup hook is running with `+o nounset`: https://github.com/NixOS/nixpkgs/commit/a6bb2ede232940a96150da7207a3ecd15eb6328c#commitcomment-36605469

In fact, when I comment out

https://github.com/NixOS/nixpkgs/blob/4c7a9a0a88ea6e8efe43661f215e0c5d32504f2e/pkgs/stdenv/generic/setup.sh#L1307-L1308

and try to build `python3.pkgs.setuptools`, it fails on trying to build `xz`:

```
/nix/store/aknix5zw9cj7hd1m3h1d6nnmncl1vkvn-multiple-outputs.sh: line 55: shareDocName: unbound variable
```

**Comment by Ericson2314** [comment]
Indeed the work isn't done! https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L1307-L1308 needs to be guarded by `IN_NIX_SHELL` so it doesn't mess up normal builds.

I should have made an issue earlier when I realized this, but didn't have the time to immediately start fixing it.

**Pull request merged**